### ISSUE TITLE
Example of 3D Charts

### ISSUE THREAD

**Issue by JannesRed** [LayTec-AG/Plotly.Blazor#44]
Hi

This is a really cool wrapper - however I am struggling to implement a 3D surface map. This there a plan to include one in the examples?

Many thanks
Jannes

**Comment by sean-mcleish** [comment]
Hi Jannes,

I'm on vacation right now and should have time again in a week. I will then upload an example.
If you have already found a solution in that time, you can create a pull-request if you want to. 

Greetings

Sean

**Comment by EuanKirkhope** [comment]
This is a heatmap, which in JS is pretty much the same as the Surface graph,  It shows how to use the XYZ cooridinates.

I have surface graph too but it doesn't compile at moment (unused code), so will need to speed some time on it.  Hopefully this gives you some pointers.

`@using Plotly.Blazor
@using PBT = Plotly.Blazor.Traces
@using PBL = Plotly.Blazor.LayoutLib

<PlotlyChart style="height: 85vh; min-height: 150px" @bind-Config="config" @bind-Layout="layout" @bind-Data="data" @ref="chart" />

@code
{
    [Parameter]
    public string Id { get; set; }

    Plotly.Blazor.PlotlyChart chart;

    Plotly.Blazor.Config config = new Plotly.Blazor.Config
    {
        Responsive = true,
        DisplayModeBar = Plotly.Blazor.ConfigLib.DisplayModeBarEnum.False
    };

    Plotly.Blazor.Layout layout = new Plotly.Blazor.Layout
    {
        Title = new PBL.Title
        {
            Text = "Heatmap",
            Font = new PBL.TitleLib.Font
            {
                Color = "#AAA"
            }
        },
        Legend = new PBL.Legend
        {
            Font = new PBL.LegendLib.Font
            {
                Color = "#AAA"
            }
        },
        XAxis = new List<PBL.XAxis>
{
            new PBL.XAxis
            {
                Title = new PBL.XAxisLib.Title
                {
                    Text = " Position (m)",
                    Font = new PBL.XAxisLib.TitleLib.Font
                    {
                        Color = "#AAA"
                    }
                }
            }
        },
        YAxis = new List<PBL.YAxis>
{
            new PBL.YAxis
            {
                Title = new PBL.YAxisLib.Title
                {
                    Text = "Time",
                    Font = new PBL.YAxisLib.TitleLib.Font
                    {
                        Color = "#AAA"
                    }
                }
            }
        },

        Scene = new List<PBL.Scene>
{
            new PBL.Scene{
                XAxis = new PBL.SceneLib.XAxis
                {
                    Title = new PBL.SceneLib.XAxisLib.Title
                    {
                        Text = "Face Position (m)",
                        Font = new PBL.SceneLib.XAxisLib.TitleLib.Font
                        {
                            Color = "#AAA"
                        }
                    }
                },
                YAxis = new PBL.SceneLib.YAxis
                {
                    Title = new PBL.SceneLib.YAxisLib.Title
                    {
                        Text = "Face Position (m)",
                        Font = new PBL.SceneLib.YAxisLib.TitleLib.Font
                        {
                            Color = "#AAA"
                        }
                    }
                },
                ZAxis = new PBL.SceneLib.ZAxis
                {
                    Title = new PBL.SceneLib.ZAxisLib.Title
                    {
                        Text = "Angle (º)",
                        Font = new PBL.SceneLib.ZAxisLib.TitleLib.Font
                        {
                            Color = "#AAA"
                        }
                    }
                },
                Camera = new Plotly.Blazor.LayoutLib.SceneLib.Camera
                {
                    Eye = new PBL.SceneLib.CameraLib.Eye
                    {
                        X = 2.5m,Y = 2m, Z = 0.5m
                    },
                    Up = new PBL.SceneLib.CameraLib.Up
                    {
                        X = 0m, Y = 0m, Z = 1m
                    },
                    Center = new PBL.SceneLib.CameraLib.Center
                    {
                        X = 0m, Y = 0m, Z = -0.3m
                    }
                },
                AspectMode = PBL.SceneLib.AspectModeEnum.Manual,
                AspectRatio = new PBL.SceneLib.AspectRatio
                {
                    X = 2m, Y = 3m, Z = 1m
                }
            }
        },

        PaperBgColor = "#262626",
        PlotBgColor = "#262626",
    };

    IList<Plotly.Blazor.ITrace> data = new List<Plotly.Blazor.ITrace>
{
        new PBT.HeatMap
        {
            Name = "SurfaceTrace",

            X = new List<object>(),
            Y = new List<object>(),
            Z = new List<object>(),
            ColorBar = new PBT.HeatMapLib.ColorBar()
            {
                OutlineColor = "#AAA",
                BorderColor = "#AAA",
                TickColor = "#AAA",
                TickFont = new PBT.HeatMapLib.ColorBarLib.TickFont
                {
                    Color = "#AAA"
                }
            }
        }
    };

    private async Task SetData()
    {
       // your get data function here
        var h = data[0] as PBT.HeatMap;
               foreach (var item in angles)
        {
             h.Y.Add(item.CreatedOn); // y is time
            h.Z.Add(item.TheHeightValue);
        }

            for (int i = 0; i < maxIdx; i++)
            {
                h.X.Add(i);
            }    
    }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            try
            {
                await SetData();
                await chart.NewPlot();
            }
            catch (Exception e)
            {

            }
        }
    }
}`

**Comment by EuanKirkhope** [comment]
I was planning to have my graph dynamically change between surface and heatmap like I do in pure javascript but with the datatypes hard typed to the graph type this doesn't work, so I've had to make two separate graphs and swap them out completely.  In js you just change graphtype from heatmap to surface.

**Comment by EuanKirkhope** [comment]
BTW, Z should be a vector x * y in length, wrong in the above example apparently :)

**Comment by EuanKirkhope** [comment]
TLDR: setting the Y axis (or any X too presumably) to a Datetime doesn't work.

I've found that in the example above if I set the line:

 h.Y.Add(item.Id); // row in database or datetime for time axis

from the db row id (integer) to timestamp (DateTime), the surface graph no longer appears, and a heatmap shows the first row repeated.

**Comment by JannesRed** [comment]
Hi, thanks! Obviously new to this package. I've managed to get a working surface map with some minor modifications to the heatmap code. 

There seems to be an XCalendar property on the Surface class which I assume has to be set to Gregorian for datetime arrays to work. Will definitely fiddle around here.

And yes, the Z axis needs to be some arrays in an array to match the data in the X and Y axes.

**Comment by sean-mcleish** [comment]
Added a surface example in v1.2.0.

Have a great week :)

**Comment by JannesRed** [comment]
Thank you, Sean :)
